site/_posts/2020-09-01-microservices-gateway.md

```diff
@@ -0,0 +1,267 @@
+const router = express.Router();
+
+router.get('/products', function(req, res, next) {
+  res.json('products: []')
+});
+
+router.get('/products/:id', function(req, res, next) {
+  res.json(`product: ${req.params.id}`)
+});
+
+module.exports = router;
+```
+
+We've created two basic routes, `/products` and `/products/:id`, so we can get a list of products and a single product. Obviously, for a real microservice, these routes would request the product information from a datastore. For now, the former is returning an empty array `[]` and the latter returns the product id requested.
+
+Try modifying your curl request to add a product ID, and notice that the response will indicate the specific ID you requested.
+
+The Product Catalog service is ready to go!
+
+## Product Inventory service
+
+Open up another terminal window and enter the `product-inventory` folder. Run `npm install` to install needed dependencies.
+
+Here's what our Product Inventory service looks like (in `routes/index.js`):
+
+```javascript
+const express = require('express');
+const router = express.Router();
+
+router.get('/branches/:id/products', function(req, res, next) {
+  const roles = req.headers.roles;
+  if (roles && roles.includes('admin')) {
+    res.json(`Products for branch #${req.params.id}`);
+  } else {
+    res.redirect(403, 'http://localhost:3000');
+    return;
+  }
+});
+
+module.exports = router;
+```
+
+In this service, we've just got one route, to get products for a specific store or branch. Notice, however, that we're allowing access (or denying it) based on the inclusion of an `admin` role in the `roles` header. The API gateway application will be responsible for passing this data to our Product Inventory service.
+
+If you were to start the service--go ahead and do so with `npm start`--and send a request to `http://localhost:3002/branches/1/products`, you should receive a 403. You can simulate a successful response by adding a `roles` header with a value of `admin`:
+
+```shell
+curl -i -H "Accept: application/json" -H "Content-Type: application/json" -H "roles: admin" http://localhost:3002/branches/1/products
+```
+
+We've got our Product Inventory service up and running, with authorization to ensure that only admins can access a list of products for a branch. Doing an authorization check at the service level allows us to granularly implement authorization.
+
+### API-level authentication
+
+A quick note on API-level authentication. We're implementing centralized authentication through the API gateway, and we want only that gateway to have access to these microservices. You could do this at the network level, or with some form of API-level authentication, like an [API Key](https://microservice-api-patterns.org/patterns/quality/qualityManagementAndGovernance/APIKey). It's a little much for us to cover in this article, but it's definitely something you'll want to implement before launching your microservices into production.
```

Comment:
the API gateway, and we ~want~ only want that gateway

Reply:
I reworked.